lib/Sema/CSSimplify.cpp

```diff
@@ -3626,18 +3626,34 @@ ConstraintSystem::matchExistentialTypes(Type type1, Type type2,
 
           if (last.is<LocatorPathElt::ApplyArgToParam>()) {
             auto proto = protoDecl->getDeclaredInterfaceType();
-            auto *fix = AllowArgumentMismatch::create(
-                *this, type1, proto, getConstraintLocator(anchor, path));
-
+            ConstraintFix *fix;
             // Impact is 2 here because there are two failures
             // 1 - missing conformance and 2 - incorrect argument type.
             //
             // This would make sure that arguments with incorrect
             // conformances are not prioritized over general argument
             // mismatches.
-            if (recordFix(fix, /*impact=*/2))
+            unsigned impact;
+            if (type1->isOptional()) {
+              auto unwrappedType = type1->lookThroughAllOptionalTypes();
+              auto matchTypeResult =
+                  matchTypes(unwrappedType, type2, ConstraintKind::Conversion,
```

Comment:
New condition is incorrect because this is happening during check that verifies that existential type represented by `type1` satisfies all protocol requirements of existential type `type2`, we cannot simply try to convert there to determine whether unwrap works or not, we need to re-run `simplifyConformsToConstraint`  (with added `TMF_ApplyingFix` flag) on unwrapped type to figure that out.

Reply:
I tried this - can you take a look and see if its what you meant?

Reply:
I think Pavel meant something like 
```cpp
auto proto = protoDecl->getDeclaredInterfaceType();
if (type1->isOptional()) {
  auto unwrappedType = type1->lookThroughAllOptionalTypes();
  auto result = simplifyConformsToConstraint(
      unwrappedType, protoDecl, kind, locator,
      subflags | TMF_ApplyingFix);
  if (result == SolutionKind::Solved) {
    auto fix = ForceOptional::create(*this, type1, proto,
                                     getConstraintLocator(locator));
    if (recordFix(fix))
      return getTypeMatchFailure(locator);
    break;
  }
}
auto fix = AllowArgumentMismatch::create(
    *this, type1, proto, getConstraintLocator(anchor, path));
// Impact is 2 here because there are two failures
// 1 - missing conformance and 2 - incorrect argument type.
//
// This would make sure that arguments with incorrect
// conformances are not prioritized over general argument
// mismatches.
if (recordFix(fix, /*impact=*/2))
  return getTypeMatchFailure(locator);

break;
```